control_plane/attachment_service/src/service.rs

```diff
@@ -178,8 +219,19 @@ impl Service {
         let mut compute_notifications = Vec::new();
 
         // Populate intent and observed states for all tenants, based on reported state on pageservers
-        let shard_count = {
+        let (shard_count, nodes) = {
             let mut locked = self.inner.write().unwrap();
+
+            // Mark nodes online if they responded to us: nodes are offline by default after a restart.
+            let mut nodes = (*locked.nodes).clone();
+            for (node_id, node) in nodes.iter_mut() {
+                if nodes_online.contains(node_id) {
+                    node.availability = NodeAvailability::Active;
+                }
+            }
+            locked.nodes = Arc::new(nodes);
+            let nodes = locked.nodes.clone();
```

Comment:
nit: Why not mutate in-place since you've already grabbed the write lock.

Reply:
The contents of an Arc are immutable - the write lock on ServiceState lets us mutate the arc itself (by assigning it to a new Arc), but not its contents.  The reason the nodes are in an Arc is so that we can cheaply clone it off for Reconciler, as a snapshot of how nodes were configured at the moment we called maybe_reconcile.  I'm not married to this overall approach, might change later to let Reconciler carry a more minimal piece of per-node information (it really just needs addresses) rather than the whole Arc<> of the map.

This is a bit clunky, and might benefit from a wrapper function for mutating nodes.

Reply:
Right, missed the fact it's an `Arc`, although it's obviously spelled out on line `232`. I agree it's a bit clunky.